### ISSUE TITLE
[EPIC] Split up assets/pools in Ocean Market

### ISSUE THREAD

**Issue by trentmc** [oceanprotocol/multi-repo-issue#59]
Fully updated spec: https://github.com/oceanprotocol/multi-repo-issue/issues/59#issuecomment-843928491

<details>
 <summary>Initial problem description</summary>
 ### What / TODO for this issue

Split apart functionality to publish / consume, from markets functionality (AMM DEX, FRE, etc). 

Pieces of functionality:
- **Publish**. Input: nothing. Output: new ERC20 datatoken, many tokens minted
- **Consume**. Input: 1.0 datatokens. Output: access to dataset/service
- **DEX**. Input: ERC20 tokens minted, >=0 bpools. Output: bpools w/ diff't amts staked & base/datatoken ratios. (Actions: stake, swap)

The first two can readily lump together; thereby being two sub-apps.

### Motivation

The grouping above orthogonalizes behavior better, to set us up for other items in 2021 roadmap. Details below. By the time we're done "NFTs" and "Better Staking", it will look a bit like the following.

- **Publish**. Input: nothing (just user entry). Output: ERC721 NFT w/ metadata.
- **Consume**. Input: 1.0 datatokens. Output: access to dataset/service
- **IDO**. Input: NFT w/ metadata. Output: ER20 datatoken, many tokens minted, bpool 
- **DEX**. Input: ERC20 tokens minted, >=0 bpools. Output: bpools w/ diff't amts staked & base/datatoken ratios. (Actions: stake, swap). Will support >1 pools per DT.
- **Info**. Input: chain info. Output: nothing (just GUI info). Be like coingecko, with two types of pages: (a) top-level summary table, with one asset per row. Has search. Click on asset, go to its detail page. (b) per asset, a detail page. Include metadata, links to all markets, price (aggregated somehow). Info page could = home page (versus an "info" tab).

However the full extent is outside the scope of _this_ issue. _This_ issue puts us in better shape without relying on NFTs being done.

### Benefits

Here are benefits once the full extent is done. A subset of them are for this issue.

- Better supports the demands of Ocean 2021 roadmap (details in later bullets)
- Support requests by enterprise engagements for other pricing methods
- Support OCEAN-backed stablecoin pricing, 3rd party marketplace token, >1 of these for a given datatoken
- More obvious / easy place to integrate "Better staking"
- Support non-$OCEAN pricing & staking (eg stablecoins, other marketplaces' tokens)
- Allow more creativity in pricing options: royalties, bonding curves, ... 
- Allow non-ERC20 datatoken pricing options
- Easier to support >1 pools, and >1 pricing options, all at once. 
- Easier to add new types of data services (beyond static url and compute to data)
- Easier to maintain: "exchange" part only deals with the economic side, nothing on publishing or consuming; and publishing / consuming part don't deal with economic side
- Better composability with other defi tools: after publish and before consume, there should be many ways to operate on ERC20 datatokens
- Gives a path to migrate existing datatokens from OceanV3 pools to 1SS/etc pools (OceanV4, whether BalancerV1 or V2)
- Easier to migrate existing datatokens from old 10-90 pools to newer ones
- More flexible support for AMM tech. Therefore less tech risk, support latest-greatest pools, leverage their liquidity mining programs. E.g. Balancer V1 & V2, Uniswap V2 & V3, Bancor, SushiSwap
- Opens the door to (eventually) pools on different chains & L2 like Secret, Moonbeam, etc. 
- More properly aligns with the actual mental model that there *can* be >1 pool and even >1 market for a given datatoken.

### Notes

After this issue is done, [market#467](https://app.zenhub.com/workspaces/ocean-main-5e5e5a97e91ce03f380fb9b7/issues/oceanprotocol/market/467) "Ocean Market supports >1 market per datatoken" will be a great follow-up towards achieving many of the benefits given above.

</details>

**Comment by kremalicious** [comment]
No listed benefit is solved through the proposed solution. Splitting up the market into different overall pages was an option in the early conception of other markets but in the end it always turned out the screens will repeat many elements anyway. So for users it becomes just more confusing.

What the current laid out spec tries to do is to put all jobs possible through the market on the same hierarchical level and it simply maps jobs-to-be-done to top-level pages. This is the basis of all confusing products and this is what we wanted to prevent. The dataset and its metadata need to stay the most important thing from which all actions are derived from.

There is of course a case to be made to value the trade part a bit more, but that could also be solved by adding one(!) new dedicated trade screen which is revolved around the datatokens and their financial data and not the dataset. On the other hand I don't see any reason to have a dedicated Consume page.

**Comment by kremalicious** [comment]
Also suggest to actually define what should be on those screens, not just limit the definition to flow of ERC-20. We build for users, not tokens. Like:

> Publish. Input: nothing. Output: new ERC20 datatoken, many tokens minted

What do you mean no input? User has to add metadata and the actual file.

> Consume. Input: 1.0 datatokens. Output: access to dataset/service

So we want a screen where users have to input their datatoken into an input field to get the dataset they want? If we list all assets on there, how would this page look different than a search result page? Why would we want to detach consume from the asset details page which is the most expected place for a user to consume?

To me only the DEX part could be worth having a dedicated page, where the primary view is organized based on the financial data of the assets. Or even simpler, just move the _Trade_ form into this so we get a global swap form. But as of right now users will only be able to swap between any datatoken - OCEAN so not a full DEX-style swap functionality. For that we would need to have a way to swap between anything, also datatoken - datatoken, so across pools.

**Comment by mihaisc** [comment]
Suggested proposal

The market is split into 2 : 
- assets (may need a better name)  https://market.oceanprotocol.com/assets
- dex https://market.oceanprotocol.com/dex

Splitting in more parts doesn't make sense from a UX standpoint. It would be cool to have dex.oceanprotocol.com and market.oceanprotocol.com but that would mean two repos and a large overhead in development. Plus users would need to switch between two websites.

The header will have a switch/toggle ( or whatever looks good) to switch between assets and market. Both sections will have the same UI but a slightly different color schema for the user to recognize at a glance.  Basically the market will have 2 landing pages based on user preferences. 

By doing this we can easily maintain all the shared components and styles and it's easier to integrate the 2 sections.

**1. Assets**

In this section there will be no concept of price. The asset costs 1 dt, that's it. 
Aquarius will not have price information. I think we need it to have the purgatory information though.
Here we can do a total refactor of the asset details page and how we show the asset. Remember, in this section the asset is number 1, the star of the show. We need to think how we emphasize what is sold.
In the asset details we have can have some graphs/stats with how many orders were made.
We can also have a small widget were we guide the user to the dex , maybe have a quick buy option from a "default" pool or a cheapest pool ( this needs more thinking)

In the search page we need to rethink the filters/sort. We will have no pricing information and we need other filters : tags , categories etc. . We need to think what makes sense for what is sold.

We can also rethink the grid where we show the asset teasers

Maybe this a good time to refactor a bit the publish form as well.

The `Buy` button in the asset details page will : check if you have 1 dt, order the asset, download the asset. It will not buy dt.

**2. DEX**

Here The Graph is the source of truth. We display all our datatokens and pools/fre. Now we have all the page at our disposal (in the past we only had the small actions zone on the right) so we can get creative.
On each datatoken we can have a quick link to the asset that is related.
Here we can stake and swap.

As well we ca fetch pools from other sources (uniswap, balancer, etc. ). We need to set a list of what is of interest. 
We can integrate these into our graph (@alexcos20 said we can do a quick integration, and this would ease the work on the front-end since we would be querying only one component) or we can query all the graphs and display them.
For the pools that are from external sources we just show minimal information, the source and a link to it.

Another suggestion would be to cleanup our terms starting with these changes. You will see that I never used did or ddo. I suggest we drop these in favor of ddo = asset and/or nft , did = key (the key/address  of the nft)

**Comment by trentmc** [comment]
Mihai - this looks like a reasonable way to go from my standpoint. +10

> Splitting in more parts doesn't make sense from a UX standpoint.

It's fine to have one "app" for both publish and consume; and one for DEX. 
We can expect one more "app" just for "IDO". But that can wait until Q3 work on [Better Staking / IDO](https://github.com/oceanprotocol/multi-repo-issue/issues/36).

> **1. Assets**
> We can also have a small widget were we guide the user to the dex , maybe have a quick buy option from a "default" pool or a cheapest pool ( this needs more thinking)

Here's another possibility. Use an affordance like CoinGecko "markets" table, which has a row for each listing (market). Each row has a link to easily go to that market.  [Here's an example on OCEAN asset](https://www.coingecko.com/en/coins/ocean-protocol#markets). 

- In our version: when the dataset is first published, there aren't any markets at first (eg no FRE, no pool, etc), so accordingly the table has no rows at first.  Then each new market gets a new row. Eg when someone creates a pool, a row gets added for that pool.  
- Each row can also have the price on _that_ market. This is great because users can quickly learn the prices; yet we don't have to come up with some heuristic to aggregate price across markets.- This approach would take us nicely towards [market#467](https://github.com/oceanprotocol/market/issues/467) "support >1 market per datatoken" and related benefits.

> The Buy button in the asset details page will : check if you have 1 dt, order the asset, download the asset. It will not buy dt.

This is a "nice to have" but not "needed", since that's what each market is for. Having it here raises the Q: if there is >1 market, where to buy? The answer quickly gets complicated, quickly, as soon as we have >1 market, trending towards aggregator functionality like 1inch. My recommendation: don't do it now, revisit after we finish this issue and [market#467](https://github.com/oceanprotocol/market/issues/467) "support >1 market per datatoken".

I didn't see comments on "Consume". How do you envision this? 

> **2. DEX**

Is the intent to hold 1 pool within a single app; or >1 pools within a single app? And, how to handle FRE?

Here's one possible way, if we have the table of markets like mentioned above:
- Each "market" has its own app instance. 
- For example, one app instance is a pool, another app instance is a second pool, and another app instance is a FRE app. (For the latter we might have a simple app like market.oceanprotocol.com/fre.

**Comment by mihaisc** [comment]
> I didn't see comments on "Consume". How do you envision this?

It's exactly this `The Buy button in the asset details page will : check if you have 1 dt, order the asset, download the asset. It will not buy dt.` . You mistook it for buying dt. I didn't use the term `consume` because it has multiple meanings. For us in the market until now consume = buy dt (from pool or exchange) ,order asset, download.

> Is the intent to hold 1 pool within a single app; or >1 pools within a single app?

Multiple, I see it as an aggregator, like i described here.

> As well we ca fetch pools from other sources (uniswap, balancer, etc. ). We need to set a list of what is of interest.
We can integrate these into our graph (@alexcos20 said we can do a quick integration, and this would ease the work on the front-end since we would be querying only one component) or we can query all the graphs and display them.
For the pools that are from external sources we just show minimal information, the source and a link to it.

> And then we'd still need to solve handling FRE and free.
For free we don't need to do anything in the dex section. You mark the asset as free and that's it.
In the DEX we can have a section where we list FRE. Simple table/list with sort and search by dt where you can buy. Nothing more needed for FRE

**Comment by mihaisc** [comment]
We need to start with defining the naming the sections of the market. Until we find better name I will go with 'Assets' and 'Dex'.

ToDo/Issue/Question format :  'Section'+'Number' . Sections: General ( G1 ..) , Assets (A1) , Dex (D1) 

**G1**: We need to move to full screen , to be actually responsive
**G2**: Header on mobile is not the best experience
**G3**: Move the search bar in the header to have a global search feature like previously discussed
**G4**: A component in the header to quickly switch between Assets and Dex . Maybe remove 'beta' tag and put something there instead. We can keep the beta banner
**G5**: Remove wallet from header. Uses too much space and will be useless. The wallet component will be shown in specific places, exactly where it is needed  (e.g. buy asset, buy datatoken, publish asset … )
**G6**: Remove chain from settings since assets from all chains will be shown (there will be filters in search)
**G7**: Look into 3box, do we need to update to ceramic? Ideally we keep our current light implementation, but we need to check how long will this be supported.
**G8**: In subgraph, for each datatoken can we have an average/aggregate price based on known/supported pools/fre?

## Assets

Pages: 

- Home
- Publish
- Search
- Asset details
- Dashboad (sort of replacing History)

**A1**: Can we have the number of orders in aquarius? If we can't then it's almost impossible to have filtering based on this since we will need to fetch and compare data from multiple subgraphs

### 1. Home

Sections :

- Most popular: short list of assets based on number of purchases (dep. **A1**)
- Recently Published ( with link to "All assets", like we have now)

Maybe we can spice up the UI, so we don't have just some grids.

Asset teaser : 

- If possible a refresh of this to have in mind new type of content (video,images,music,etc.)
- Maybe have a possibility to have a preview image
- Bigger?

**A2**: Does it make sense to have data type ( Image (png,jpeg), audio (flac, mp3, ...), video (mp4, ...), data (json, csv, ...) ) ? This is data introduced by the publisher. Using this data we may pick a default preview image if none added by user.

-  Visible medatada (not necessarily represented by text, it can be represented by icons/images like we do with type and services):
  - Title
  - Price (based on G8)
  - Type : data/algo
  - Service : compute/access
  - Created/Uploaded date
  - Datatoken name
  - Publisher name/address
  - Data type? image, audio, ... (dep. **A2**)
  - Network/Chain ? Do we need this on the teaser? Nice to have if we have room. If think the main goal is to get the user on the details page (and not discourage them if the asset is on a different chain, basically the asset teaser should actually tease the user not show all the data)

### 2. Publish

We need to design a proper wizard (maybe get a wizard component if it's light enough). The last step of the wizard should be the price creation 

### 3. Search

PRIO ( this can be done in parallel, doesn't have to be included in this major epic) : we need a useful and relevant search. The current search is inaccurate and useless. Quick test, the asset is not on the first page:

![image](https://user-images.githubusercontent.com/26000280/118790511-8162a000-b89e-11eb-8a1d-2e4efcf13796.png)
![image](https://user-images.githubusercontent.com/26000280/118790526-83c4fa00-b89e-11eb-8b9f-2f0804a673fc.png)


Filters:

- Type: data/algorithm
- Service: compute/access
- Data Type (dep. **A2**)

Sort:

- Published
- Popularity (dep **A1**)

### 4. Asset details

Left part can stay as it is (the one with the details).

We need to rethink the actions

![image](https://user-images.githubusercontent.com/26000280/118790575-92abac80-b89e-11eb-8e68-5b06c2b70d93.png)


**A4**: Do we want to explore the possibility where we support having multiple services?



### 5. Dashboard

We need to get creative here. What i think would be cool here:

- Sales (fancy charts & stats : per item, overall)
- Buy history
- Compute stats and results



## DEX

Pages:

- Home/Search
- Details
- Dashboard

### 1. Home/Search

List of all pools and fre. I think a table might be a better choice here.

Filter by :

- tokens
- price
- tvl

Data to be displayed for each item :

- tokens
- tvl
- spot price
- connected asset

### 2. Details

Here we add all the stats and actions that we removed from asset details (even more).

### 3. Dashboard

View all pool shares, view stats (how much did your share value increase/decrease)
View all tx
We can do multiple things here in multiple steps, no need to get stuck on it.




Next steps : 

- Decide what features stay 
- Create epics/issues

**Comment by kremalicious** [comment]
## Menu

- Assets
- Pools
- Publish
- Account

## Routes and internal nomenclatura:

- `/`

### Assets

- `/assets/`: Assets index
- `/asset/did:op:xxxxxxxx`: Asset details page

### Pools

- `/pools/`: Pools index
- `/pool/0xxxxxxxxx`: Pool details page

### Account

- `/account/0xxxxxxx`: Account details page

### Publish

- `/publish`: Publish page

### Search

- `/search`: Search page

**Comment by mihaisc** [comment]
Suggested feature rollout: 

Beta (probably 2-3 sprints ) : 
- new home page, assets page, pools page, pool details page (this will probably go trough multiple iterations)
- new search page ( all functionality may not be possible at this point)

We can use netlify split testing with a ratio of 6:4 or 7:3 ( a part of users get the main branch , the rest get the new branch)

Final (2-4 sprints ) : 
- add new publish
- add new asset details
